src/email/email.service.ts

```diff
@@ -0,0 +1,54 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { MailerService } from '@nestjs-modules/mailer';
+import { TokensService } from 'src/tokens/tokens.service';
+import { User } from 'src/users/entities/user.entity';
+import { MailSubject, MailTemplate } from 'src/utils/mail/mail-details';
+
+import { RegistrationLinks, TokenData, TokenOptions } from '../types';
+
+@Injectable()
+export class EmailService {
+  constructor(
+    private readonly mailerService: MailerService,
+    private readonly tokensService: TokensService,
+    private readonly configService: ConfigService,
+  ) {}
+
+  private async createUrls(users: User[], tokenType: TokenOptions): Promise<RegistrationLinks[]> {
+    const registrationLinks: RegistrationLinks[] = [];
+    const promises = users.map(async user => {
+      const tokenData: TokenData = await this.tokensService.createToken(user, tokenType);
+      const { token } = tokenData;
+      const url = `http://localhost:3000/api/auth/register/${user.id}/${token}`;
+      registrationLinks.push({
+        email: user.email,
+        url,
+      });
+    });
+    await Promise.all(promises);
+    Logger.log('users', users, 'tokenType', tokenType);
+    return registrationLinks;
+  }
+
+  async sendRegisterConfirmation(email: string[]): Promise<void> {
```

Comment:
Kontynuując poprzedni komentarz: I tu się tylko odpalasz ładną metodę generującą tablicę {link, email} i robisz ekspresową wysyłkę

Edit: ale to też można ładnie rozbić :)
w `EmailService` robisz tylko metodę do wysłania pojedynczej wiadomości z linkiem aktywacyjnym (albo jeszcze bardziej uogólniajac - z linkiem). W parametrach metody podajesz emali, link i rodzai template'u. 
Wtedy mamy już metodę do wysłania i linku aktywacyjnego (dla grupy studentów/pojedynczego hr), jak również do wysłania linku ze zmianą hasła (manipulujesz tu szablonem mailowym).

Natomiast w `StudentsService` tworzysz metodę np. `sendBulkActivationEmails` w której spinasz to wszystko co do tej pory zrobiłeś i masz ładną obsługę wysłania linków zbiorczych :)